src/lender/tranche/operator/proportional.sol

```diff
@@ -0,0 +1,170 @@
+        if(what == "supplyAllowed") {
+            supplyAllowed = supplyAllowed_;
+        }
+    }
+
+    function updateReturned(uint currencyReturned, uint principalReturned) public auth {
+        totalCurrencyReturned  = safeAdd(totalCurrencyReturned, currencyReturned);
+        totalPrincipalReturned = safeAdd(totalPrincipalReturned, principalReturned);
+    }
+
+    /// defines the max amount of currency for supply
+    function approve(address usr, uint currencyAmount) external auth {
+        supplyMaximum[usr] = currencyAmount;
+    }
+
+    /// only approved investors can supply and approved
+    function supply(uint currencyAmount) external note {
+        require(supplyAllowed);
+
+        require(safeSub(supplyMaximum[msg.sender], tokenReceived[msg.sender]) >= currencyAmount, "not-enough-currency");
+        tokenReceived[msg.sender] = safeAdd(tokenReceived[msg.sender], currencyAmount);
+
+        require(assessor.supplyApprove(address(tranche), currencyAmount), "supply-not-approved");
+
+        // pre-defined tokenPrice of ONE
+        uint tokenAmount = currencyAmount;
+
+        tranche.supply(msg.sender, currencyAmount, tokenAmount);
+
+        totalPrincipal = safeAdd(totalPrincipal, currencyAmount);
+
+        distributor.balance();
+    }
+
+    /// redeem is proportional allowed
+    function redeem(uint tokenAmount) external note {
+        distributor.balance();
+        (uint currencyAmount, uint currencyRedeemed_,uint  principalRedeemed_) = calcRedeemCurrencyAmount(msg.sender, tokenAmount);
+        currencyRedeemed[msg.sender] = currencyRedeemed_;
+        principalRedeemed[msg.sender] = principalRedeemed_;
+        require(assessor.redeemApprove(address(tranche), currencyAmount), "redeem-not-approved");
+        tokenRedeemed[msg.sender] = safeAdd(tokenRedeemed[msg.sender], tokenAmount);
+        tranche.redeem(msg.sender, currencyAmount, tokenAmount);
+    }
+
+    /// calculates the current max amount of tokens a user can redeem
+    /// the max amount of token depends on the total principal returned
+    /// and previous redeem actions of the user
+    function calcMaxRedeemToken(address usr) public view returns(uint) {
+        if (supplyAllowed) {
+            return 0;
+        }
+        // assumes a initial token price of ONE
+        return safeSub(rmul(rdiv(totalPrincipalReturned, totalPrincipal), tokenReceived[usr]), tokenRedeemed[usr]);
+    }
+
+    /// calculates the amount of currency a user can redeem for a specific token amount
+    /// the used token price for the conversion can be different among users depending on their
+    /// redeem history.
+    function calcRedeemCurrencyAmount(address usr, uint tokenAmount) public view returns(uint, uint, uint) {
+        // solidity gas-optimized calculation avoiding local variable if possible
+
+        uint maxTokenAmount = calcMaxRedeemToken(usr);
+        require(tokenAmount <= maxTokenAmount, "tokenAmount higher than maximum");
```

Comment:
@manuelpolzhofer I think we should make this method just use the maximum allowed in case tokenAmount > maxTokenAmount instead of a revert. Will make it simpler in the UI.

Reply:
okay i will add